### PR TITLE
Add color argument to button mixins

### LINKED ISSUE #28899: Should button-variant function allow a manually-specified color?
In [Bootstrap 3, the button-variant function/mixin](https://github.com/twbs/bootstrap-sass/blob/master/assets/stylesheets/bootstrap/mixins/_buttons.scss#L6) had this signature:

    @mixin button-variant($color, $background, $border)

You specified the (text) color, background color, and border. Other values (such as focus/hover colors) were computed automatically derived from these. 

In [Bootstrap 4, the button-variant function/mixi](https://github.com/twbs/bootstrap/blob/2ac2a5a91e2958a367c46a61d05162f640457db5/scss/mixins/_buttons.scss#L6) has this signature instead:

    @mixin button-variant($background, $border, $hover-background: darken($background, 7.5%), $hover-border: darken($border, 10%), $active-background: darken($background, 10%), $active-border: darken($border, 12.5%)) {

You specify, at least, the background and border color. All other values are computed derived from those. You can *optionally* specify your hover and active border/backgrounds instead. The [built-in uses of button-variant](https://github.com/twbs/bootstrap/blob/2ac2a5a91e2958a367c46a61d05162f640457db5/scss/_buttons.scss#L61) only specify $background and $border (and actually specify $border to be the same value as $background), and allow the default computation for the rest. 

There is no way to manually specify the actual (text/foreground) `color` as an option though -- for default, hover, or active. You are stuck with the computed/derived value based on [contrasting colors with specified/default backgrounds](https://github.com/twbs/bootstrap/blob/2ac2a5a91e2958a367c46a61d05162f640457db5/scss/mixins/_buttons.scss#L7). 

If you wanted to manually override this to something that worked better, for your particular palette or branding stylesheet, than the automatically computed value, you could do it by applying button-variant mixin and then providing CSS overrides -- but the need to deal with all the pseudo-selectors and nesting (which can be [complex and non-obvious](https://github.com/twbs/bootstrap/blob/2ac2a5a91e2958a367c46a61d05162f640457db5/scss/mixins/_buttons.scss#L43) ) makes this kind of challenging to do in a reliable way. 

Should button-variant function allow optional arguments with manually specified `color` values? Which if not provided, would still default to the automatically computed value with `color-yiq` and relevant background color.

### PR DESCRIPTION
**Pull request by jeremyvii**
Fix #28899 by adding optional color arguments, for each button state, to `button-variant()` and `button-outline-variant()`. The color contrast function was moved to the default values of these mixins to reduce the need for conditions inside the mixin body.

To do:
- [x] Add documentation

### CODE REVIEW

**Comment by MartijnCuppens** [comment]
This breaking change should be documented in the migration guide (probably under the Sass section).

File: scss/mixins/_buttons.scss [hunk 1]
```diff
@@ -3,35 +3,35 @@
 // Easily pump out default styles, as well as :hover, :focus, :active,
 // and disabled options for all buttons
 
-@mixin button-variant($background, $border, $hover-background: darken($background, 7.5%), $hover-border: darken($border, 10%), $active-background: darken($background, 10%), $active-border: darken($border, 12.5%)) {
-  color: color-yiq($background);
+@mixin button-variant($background, $border, $color: color-yiq($background), $hover-background: darken($background, 7.5%), $hover-border: darken($border, 10%),  $hover-color: color-yiq($hover-background), $active-background: darken($background, 10%), $active-border: darken($border, 12.5%), $active-color: color-yiq($active-background)) {
```
Comment (mdo) [inline scss/mixins/_buttons.scss]: Could we start to reformat our mixins with all these parameters? Starting to feel difficult to read :).

```suggestion
@mixin button-variant(
  $background,
  $border, $color: color-yiq($background),
  $hover-background: darken($background, 7.5%),
  $hover-border: darken($border, 10%), 
  $hover-color: color-yiq($hover-background),
  $active-background: darken($background, 10%),
  $active-border: darken($border, 12.5%),
  $active-color: color-yiq($active-background)
) {
```

Comment (jeremyvii) [reply scss/mixins/_buttons.scss]: I agree, that would be much easier to read. Was the `$color` argument, on line 8, intentional on the same line as `$border`? Or should it be moved to it's own newline?

Comment (MartijnCuppens) [reply scss/mixins/_buttons.scss]: > Or should it be moved to it's own newline?

Yeah, it should (I've updated @mdo 's comment). We should also do this with the outlined buttons.

**Review by mdo — APPROVED** [review]

**Comment by MartijnCuppens** [comment]
The parameters in the button mixin have changed, this should be documented in migration guide.

**Review by XhmikosR — DISMISSED** [review]
Please address MartijnCuppens's comment

**Comment by jeremyvii** [comment]
I updated the migration guide to include the changes to `button-outline-variant()` along with what I had added for `button-variant()`. Was this all I needed to document, or was there somewhere else I needed to address?

**Comment by XhmikosR** [comment]
@MartijnCuppens LGTY now?

**Review by MartijnCuppens — APPROVED** [review]
Looks fine, thanks!

**Pull request merged**